scripts/launch-xrd

```diff
@@ -135,9 +135,20 @@ _runcmd() {
 #   Arg 1: image name
 _get_platform() {
     local platform
-    platform="$($CONTAINER_EXE inspect "$1" \
-        --format='{{index .Config.Labels "com.cisco.ios-xr.platform"}}' \
-        2>/dev/null)" || true
+
+    check_img_cmd=("$CONTAINER_EXE" "inspect" "$1" \
+            "--format={{index .Config.Labels \"com.cisco.ios-xr.platform\"}}")
+
+    if ! platform=$("${check_img_cmd[@]}" 2>/dev/null); then
+        echo "Specified image is not loaded in the local registry, attempting to pull..." >&2
+        if "$CONTAINER_EXE" pull "$1" >&2; then
+            platform=$("${check_img_cmd[@]}" 2>/dev/null)
+        else
+            echo "Unrecognised image '$1' - should be either an image loaded " \
+                 "in the local $CONTAINER_MGR registry or an image that can be pulled." >&2
+            return 1
+        fi
+    fi
```

Comment:
I don't think this logic should go in `_get_platform()`, it's a bit of a strange side effect for a function that "gets the platform from the container image labels" to pull an image. I think the place that checks whether platform was specified should do the logic of pulling the image if required.

Reply:
This is because if the platform is provided, `launch-xrd` doesnt need to get the platform. docker can take care of it.

Reply:
I know but that branching logic can still be done in the place I suggested, just like this function is only called if it's needed.

Reply:
Sorry, deleted my comment by mistake. for the record, I had said, if platform is given, `launch-xrd` doesnt need to pull the image.

Reply:
This would result in code replication. Since to check an image, you would have to inspect it. So you would have to inspect twice.

Reply:
Can rename the `_get_platform` function to `_check_image` or something?

Reply:
Here's what the flow should be:
- If platform not specified (line 303):
  - If image *not loaded*, then try to pull it, giving useful error if the pull fails or the image is still not loaded
  - Call `_get_platform` on the now-loaded image (which does the 'inspect' command)

What we currently do, which needs changing, is that if the platform isn't specified we try inspecting, and if that fails we pull... But this means if 'ubuntu' is given we'll always try pulling even if it's already loaded, and in some cases this could lead to the loaded image actually being updated (this is a side effect that should never be caused by us!).

Reply:
I was trying to say to check if an image is loaded or not. We are inspecting it. That would result in duplication of code. Even if ubuntu is loaded, `check_img_cmd` would still succeed, `platform` would be "". So there wouldn't be an update.

```
docker pull ubuntu
> _get_platform ubuntu
Error: Could not determine platform from image 'ubuntu' labels.
Use '--platform' argument instead.
```

Reply:
I agree with Lewis that the `_get_platform` function should just be inspecting and validating a docker image, and I agree with you that we don't want to repeating "checking an image is loaded" and "inspecting the platform".

I think remove image pulling from the `_get_platform` function and add another function wrapping get platform that pulls `$1` first. Something like:
```bash
if [[ ! $PLATFORM ]]; then
    PLATFORM="$(_get_platform "$IMAGE")" ||
        PLATFORM="$(_get_platform_from_registry)" || exit 1
fi
```
or with `if` statements if you prefer reading that.

Reply:
Or alternatively - something more in line with what Lewis was suggesting
```bash
if [[ ! $PLATFORM ]]; then
    if ! docker inspect "$IMAGE" &>/dev/null; then
        _pull_image "$IMAGE" || exit 1
    fi
    PLATFORM="$(_get_platform "$IMAGE")" || exit 1
fi
```
Where "pull image" is a new function that has all that nice error output in this change added to it and "get platform" is a simple inspect function.